src/lib/ThumbnailsSidebar.js

```diff
@@ -0,0 +1,186 @@
+import VirtualScroller from './VirtualScroller';
+
+const DEFAULT_THUMBNAILS_SIDEBAR_WIDTH = 150;
+const THUMBNAIL_WIDTH_MAX = 210;
+const THUMBNAIL_MARGIN = 15;
+
+class ThumbnailsSidebar {
+    /** @property {HTMLElement} - The anchor element for this ThumbnailsSidebar */
+    anchorEl;
+
+    /** @property {PDfViewer} - The PDFJS viewer instance */
+    pdfViewer;
+
+    /** @property {Object} - Cache for the thumbnail image elements */
+    thumbnailImageCache;
+
+    /**
+     * [constructor]
+     *
+     * @param {HTMLElement} element - the HTMLElement that will anchor the thumbnail sidebar
+     * @param {PDFViewer} pdfViewer - the PDFJS viewer
+     */
+    constructor(element, pdfViewer) {
+        this.anchorEl = element;
+        this.pdfViewer = pdfViewer;
+        this.thumbnailImageCache = {};
+
+        this.createPlaceholderThumbnail = this.createPlaceholderThumbnail.bind(this);
+        this.requestThumbnailImage = this.requestThumbnailImage.bind(this);
+        this.createThumbnailImage = this.createThumbnailImage.bind(this);
+        this.generateThumbnailImages = this.generateThumbnailImages.bind(this);
+    }
+
+    /**
+     * Destroys the thumbnails sidebar
+     *
+     * @return {void}
+     */
+    destroy() {
+        if (this.virtualScroller) {
+            this.virtualScroller.destroy();
+            this.virtualScroller = null;
+        }
+
+        this.thumbnailImageCache = null;
+        this.pdfViewer = null;
+    }
+
+    /**
+     * Initializes the Thumbnails Sidebar
+     *
+     * @return {void}
+     */
+    init() {
+        this.virtualScroller = new VirtualScroller(this.anchorEl);
+
+        // Get the first page of the document, and use its dimensions
+        // to set the thumbnails size of the thumbnails sidebar
+        this.pdfViewer.pdfDocument.getPage(1).then((page) => {
+            const { width, height } = page.getViewport(1);
+
+            // If the dimensions of the page are invalid then don't proceed further
+            if (!(isFinite(width) && width > 0 && isFinite(height) && height > 0)) {
+                console.error('Page dimensions invalid when initializing the thumbnails sidebar');
+                return;
+            }
+
+            // Amount to scale down from fullsize to thumbnail size
+            this.scale = DEFAULT_THUMBNAILS_SIDEBAR_WIDTH / width;
+            // Width : Height ratio of the page
+            this.pageRatio = width / height;
+            const scaledViewport = page.getViewport(this.scale);
+
+            this.virtualScroller.init({
+                totalItems: this.pdfViewer.pagesCount,
+                itemHeight: scaledViewport.height,
+                containerHeight: this.anchorEl.parentNode.clientHeight,
+                margin: THUMBNAIL_MARGIN,
+                renderItemFn: this.createPlaceholderThumbnail,
+                onScrollEnd: this.generateThumbnailImages,
+                onInit: this.generateThumbnailImages
```

Comment:
Minor, but the name makes it seem like this should be passed into the constructor.